core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java

```diff
@@ -0,0 +1,255 @@
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.routing.*;
+import org.elasticsearch.common.collect.ImmutableOpenIntMap;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.common.util.concurrent.CountDown;
+import org.elasticsearch.gateway.AsyncShardFetch;
+import org.elasticsearch.gateway.TransportNodesListGatewayStartedShards;
+import org.elasticsearch.gateway.TransportNodesListGatewayStartedShards.NodeGatewayStartedShards;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import static org.elasticsearch.action.admin.indices.shards.IndicesShardsStoresRequest.*;
+
+/**
+ * Transport action that reads the cluster state for shards with the requested criteria (see {@link Status}) of specific indices
+ * and fetches shard store information from all the nodes using {@link TransportNodesListGatewayStartedShards}
+ */
+public class TransportIndicesShardsStoresAction extends TransportMasterNodeReadAction<IndicesShardsStoresRequest, IndicesShardsStoresResponse> {
+
+    private final TransportNodesListGatewayStartedShards listShardStoresInfo;
+
+    @Inject
+    public TransportIndicesShardsStoresAction(Settings settings, TransportService transportService, ClusterService clusterService, ThreadPool threadPool, ActionFilters actionFilters,
+                                              TransportNodesListGatewayStartedShards listShardStoresInfo) {
+        super(settings, IndicesShardsStoresAction.NAME, transportService, clusterService, threadPool, actionFilters, IndicesShardsStoresRequest.class);
+        this.listShardStoresInfo = listShardStoresInfo;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected IndicesShardsStoresResponse newResponse() {
+        return new IndicesShardsStoresResponse();
+    }
+
+    @Override
+    protected void masterOperation(IndicesShardsStoresRequest request, ClusterState state, ActionListener<IndicesShardsStoresResponse> listener) {
+        final Set<ShardRoutingState> routingStates = new HashSet<>(request.shardStatuses().length);
+        for (Status status : request.shardStatuses()) {
+            switch (status) {
+                case GREEN:
+                case ALL:
+                    routingStates.addAll(Arrays.asList(ShardRoutingState.values()));
+                    break;
+                case YELLOW:
+                case RED:
```

Comment:
I don't think this is correct?  for yellow we want to have the primary assigned and at least one replica unassigned . for red, the primary should not be assigned..